### ISSUE TITLE
Infer the endLocation of a problem from the reported node

### ISSUE THREAD

**Issue by not-an-aardvark** [eslint/eslint#8004]
## Problem

Currently, the `context.report` API allows rules to specify a start and end location for problems:

```js
context.report({
  loc: {
    start: {
      line: 1,
      column: 3
    },
    end: {
      line: 1,
      column: 5
    }
  },
  message: 'Detected a problem.'
});
```

This was added to address https://github.com/eslint/eslint/issues/3307, and it's very useful because it allows editor integrations to highlight an entire range.

However, as currently implemented this API is tedious to use, because it only works if a report includes a`loc` key. As a result, most rules don't use this API, because they report a *node* instead of an explicit location. (When a node is reported, only the start location of the node appears in the resulting message -- the end location does not appear.)

## Proposal

When a node is reported without a location, the end location of the node should be used as the end location of the report. (The start location of the node is already used for the start location of the report.)

For example, given the following node:

```json
{
  "type": "Identifier",
  "start": 0,
  "end": 3,
  "name": "foo"
}
```

Suppose a rule uses

```js
context.report({ node, message: 'report message' });
```

The resulting problem message *currently* looks like this:

```json
{
  "ruleId": "some-rule",
  "severity": 2,
  "message": "report message",
  "line": 1,
  "column": 1,
  "nodeType": "Identifier",
  "source": "foo"
}
```

With this change, the resulting problem message would look like this:

```diff
{
  "ruleId": "some-rule",
  "severity": 2,
  "message": "report message",
  "line": 1,
  "column": 1,
+ "endLine": 1,
+ "endColumn": 4,
  "nodeType": "Identifier",
  "source": "foo"
}
```

**Comment by ilyavolodin** [comment]
I was under impression that we are already doing this. But if we are not, then I definitely support this.

**Comment by mysticatea** [comment]
I avoided it: #6640

Because some rules in core and plugin ecosystem have sometimes used big node (Program, FunctionDeclaration, Property, .*Statement, ...). I'm afraid that it fills editors with red.

**Comment by platinumazure** [comment]
Maybe we could implement it and then try to fix rules using large nodes as reports come in?

**Comment by not-an-aardvark** [comment]
Personally I think the issue with filling a screen is an editor integration problem, not an ESLint problem. I'm fine with waiting until a major release, but if we do that then we might also want to give some warning so that editors can display the best possible UI.

**Comment by mysticatea** [comment]
For example:

```js
return {
    Program(node) {
        context.report({ node, message: 'report message' });
    }
}
```

In this case, currently the report is at the head of the source code, but the report spans whole the source code after this change. It's a breaking change that plugin owners need to modify their plugins.

**Comment by mysticatea** [comment]
I don't oppose this if it happens on the next major release.

**Comment by not-an-aardvark** [comment]
Since it looks like we're probably going ahead with this change, we should probably notify editor integrations about this in case they want to change their behavior ahead of time. I pulled the list from the [integrations page](http://eslint.org/docs/user-guide/integrations):

- [ ] [linter-eslint](https://github.com/AtomLinter/linter-eslint)
- [ ] [vscode-eslint](https://github.com/Microsoft/vscode-eslint)
- [ ] [SublimeLinter-eslint](https://github.com/roadhump/SublimeLinter-eslint)
- [ ] [eslint-plugin (IntelliJ)](https://github.com/idok/eslint-plugin
- [ ] [TextMate 2 bundle](https://github.com/natesilva/javascript-eslint.tmbundle)
- [ ] [tern-eslint](https://github.com/angelozerr/tern-eslint)
- [ ] [Orion/ESLint](https://wiki.eclipse.org/index.php?title=Orion/ESLint&action=edit) (not sure where the issue tracker is)
- [ ] [Flycheck](https://github.com/flycheck/flycheck) (has an eslint setting)
- [ ] [Vim](https://github.com/vim-syntastic/syntastic/tree/master/syntax_checkers/javascript)

I haven't used most of these integrations, so I'm not sure how many of them are maintained. Also, there are probably some of integrations that wouldn't be affected by this change anyway.

**Comment by mysticatea** [comment]
At least, `vscode-eslint` and `atom-linter` are using `endLine` and `endColumn` to show reports.

I think the announcement is needed for plugin's authors rather than editor integrations because this is a breaking change of `RuleContext` API and the form of reports are not changed. This change does not affect to editors, but affects to plugins.

**Comment by kaicataldo** [comment]
To add to the list of Vim editor integrations, [neomake](https://github.com/neomake/neomake) is also quite popular alongside Syntastic.

**Comment by not-an-aardvark** [comment]
I'll add this to the TSC agenda since it's a change to core that could break things.

**TSC Summary**: As currently implemented, the `endLine` and `endColumn` report API is tedious to use, because it only works if a report includes a `loc` key. As a result, most rules don't use this API, because they report a node instead of an explicit location. We could update core to infer the `endLine` and `endColumn` of a reported node, but this would be a breaking change because it could cause a poor user experience in some existing editor integrations (e.g. if hundreds of lines are highlighted because a large node is reported).

**TSC Question:** Should we make this change?

**Comment by not-an-aardvark** [comment]
In today's TSC meeting, the TSC accepted this issue.
